5 PRs in one (for discussion)...

clehene opened the pull request:
I'd like to discuss these changes on the PR and see how much we can integrate :)

* Add support for RHEL7/CentOS7 
* Minor refactoring to remove (some) duplication in `.pp` files 
* Added / changed some configs/defaults
* Forced yum cache expiry to actually get updates
* Changed configuration templates so that config files could be sourced directly by service units

deric commented on templates/master.erb:
```diff
@@ -1,29 +1,13 @@
+MESOS_cluster="<%= @cluster %>"
+MESOS_log_dir="<%= @log_dir %>"
+MESOS_logbufsecs=1
+MESOS_logging_level="INFO"
```
I don't like default that you can't change. Though, in this case you could override this value in `@env_var` with same key.

clehene replied:
@deric I hope I understand your comment correctly.
As this is a template it could be used to generate defaults or other files.
I'm suggesting it should be based on the environment variables and, hence could be sourced (e.g. by systemd). 

The current one relies on variables that are replaced with environment variables by a script that is not versioned with mesos (not even source if it's available).

IMHO there are too many ways to configure things (this, defaults, configs, files, env vars and cli args) and this could be confusing and also unneeded in Puppet managed environments.

deric replied:
We should decide whether we want to configure Mesos via properties (`/etc/mesos-master/*` and `/etc/mesos-slave/*`) or by using ENV `MESOS_{variables}`. I think supporting both at the same time is very confusing. CLI arguments are out of questions, we expect people run Mesos as service using systemd, sysvinit or whatever else. When someone checks Mesos machine configuration will he look into `/etc/default/mesos-slave` or `/etc/mesos-slave`? Each Mesos version adds new parameters, it's problematic to add new variables each time this happens.

deric commented on templates/service/systemd/mesos.service.erb:
```diff
@@ -0,0 +1,24 @@
+Wants=network.target
+
+[Service]
+ExecStart=/usr/sbin/mesos-<%= @name %> <% if @name=='slave' -%>-containerizers=docker,mesos<% end -%>
```
here you have hard-coded `containerizers`, that's not good

clehene replied:
Do you think we need a list that we should generate from in here?

deric replied:
It should be a parameter (string) or maybe an array. 
Anyway you're trying to replace file that is already [part of a binary package](https://github.com/mesosphere/mesos-deb-packaging/blob/master/systemd/master.systemd) and bypassing the `mesos-init-wrapper` which would normally read all the properties from `/etc/mesos-master`. Any good reason why to do that?

clehene replied:
The script you refer to is not part of Mesos. Unless we'd want this module to be dependent on a 3rd party packaging (note that I'm deploying RPMs) we shouldn't rely on logic specific to it (nor to anything specific to a particular package for that matter as I think the puppet module should be agnostic to that).
if the env variables created based on that script logic make sense, perhaps we could try add it to the Mesos tree, but IMHO it offers yet another point of configuration (the 4th after cli, env, mesos conf files) and not sure it makes sense for it to exist.

deric commented on manifests/init.pp:
```diff
@@ -27,14 +27,15 @@
-  $listen_address = undef,
-  $repo           = undef,
+  $listen_address = '0.0.0.0',
+  $repo           = 'mesosphere',
```
`mesossphere` repo might not be available for all OS, this could be a breaking change for existing installations

clehene replied:
We seem to only support RedHat and Debian and both treat `'mesosphere'` value
https://github.com/deric/puppet-mesos/blob/master/manifests/repo.pp#L20-L21
https://github.com/deric/puppet-mesos/blob/master/manifests/repo.pp#L38-L39

Not sure if this should break anything, but I could just inject it from our config, if you think it's better. 
LMK

deric replied:
Right now Mesosphere is best distribution channel for binary Mesos packages, I guess we can leave it as default.

deric commented on manifests/service.pp:
```diff
@@ -13,6 +13,24 @@
+    }
+  }
+  $zk = $mesos::zk
+  $quorum = $mesos::quorum
```
why do you have `$quorum` variable in here?

deric commented on templates/master.erb:
```diff
@@ -1,29 +1,13 @@
-# IP address to listen on (in case that you start master
-# on this node, it will bind to this address)
-<% if @listen_address -%>
-IP="<%= @listen_address %>"
```
In case that `@listen_address` is `undefined` we leave up to Mesos to which address it will bind. In most cases it should be same as enforcing `0.0.0.0`, though it might break some installations...

adragomir replied:
@deric According to the documentation, Mesos takes either MESOS_IP, or MESOS_IP_DISCOVERY_COMMAND. If both are set, Mesos will throw an error. This is an attempt to be able to pass any of them here (we pass in MESOS_IP_DISCOVERY_COMMAND as extra parameters).

deric replied:
Yeah, but you hardcoded `MESOS_ip` earlier, so it won't be possible to use discovery command.

deric commented:
Generally I like the changes, don't be mistaken by too many comments. Definitely we have to fix the tests and try not to break current behaviour. If the changes would come in separate PR, it would be much easier to merge them. Anyway, we should probably bump the version at least to `0.6` or `1.0`.

clehene commented:
Thanks @deric I'll address the comments and make multiple PRs

felixb commented:
Is there any update on this?
I just wanted to implement the same thing. Native systemd support would be very nice.

deric commented:
It would be nice to separate this into multiple PR, or create at least an issue for systemd support.

felixb commented:
@clehene: are you working on this?

clehene commented:
@felixb @deric yes, we use this internally and had very recently made a few more changes we'd like to push upstream.
This said, @deric please see my questions above

ping @adragomir

clehene commented:
We've rebased on top of master, need to cleanup patches and will make new PRs

felixb commented:
any updates?

clehene commented:
@felixb I rebased everything we have on top of current master and will make separate PRs for each one CentOS7 will come as a single commit/PR but I recommend we discuss the whole group. 

@deric there are a few questions which you still haven't answered. I think it's worth discussing them, can you please take a look over the comments on the PR. Thanks!

deric commented on manifests/init.pp:
```diff
@@ -27,14 +27,15 @@
   $conf_dir       = '/etc/mesos',
   # e.g. zk://localhost:2181/mesos
   $zookeeper      = '',
+  $quorum         = 1,
```
Until now `quorum` variable could have been set via options:
```puppet
  options => {
    quorum   => 4
  }
```
But which one should take precedence? We don't want to break existing installations, so from this point of view it should be overridden by options value, if defined.

blsmth commented:
any status on this?    I'm currently trying to use this module on CentOS7 since puppet forge advertises it as compatible, but i'm slowly finding it's not fully.   some strange issues that may be resolved by this fix

deric commented:
@blsmth What kind of issues you're having? Could you create a separate issue, so that we can address then individually?

blsmth commented:
thanks @deric   see [this issue](https://github.com/deric/puppet-mesos/issues/66)

clehene commented:
@felixb @blsmth @deric ping  rebased to master and updated 
Sorry it took forever, but every time I wanted to push this and rebased it got into another round of conflicts with master...

deric commented:
@clehene cool, looks much better! What's the reasoning for introducing `mesos::common::mkdir_p`? Can't we just use:
```puppet
file{'/some/path':
  ensure  => present,
  recurse => true,
} 
``` 
otherwise we have to handle various duplicate declaration conflicts. The exec should contain at least `creates`:
```puppet
exec { "mkdir_p-${name}":
  creates => $name,
  command => "mkdir -p ${name}",
  unless  => "test -d ${name}",
  path    => '/bin:/usr/bin',
}
```
But it would be better if we could manage it just with default `file` resource.

clehene commented:
@deric `ensure => directory`, `recurse => true` will only affect ownership (and probably other properties) but won't create the parent directories. It was decided that that's too complicated (https://projects.puppetlabs.com/issues/86) 
Note that `mkdir_p` should have a reference to the original author (did I lose it during rebase?) https://github.com/ghoneycutt/puppet-module-common/blob/master/manifests/mkdir_p.pp
I'll add `creates`

clehene commented:
The commit "Changed templates to set `MESOS_` env vars" should have been squashed (at least partially) into the main RHEL7 support  as it's required because SystemD `EnvironmentFile` doesn't work with `export`

clehene commented:
Question: this is currently sourcing `EnvironmentFile=/etc/default/mesos-<%= @name %>`
If you'd like to maintain the previous functionality based on env vars (not sure it's needed), we could probably have both files created and the template could add "export" optionally, based on a variable or something.

felixb commented:
Imho, the config should live in `/etc/sysconfig/` on centos.

clehene commented:
thanks @felixb for pointing that. I'll fix it.

clehene commented:
Quick note. I'm seeing some issues (`yum list` doesn't return anything because the repo is not being installed.) I'll dig into that and fix it. It's likely a missing dependency.

deric commented:
@clehene What is the reason for removing all:
```puppet
  owner  => $owner,
  group  => $group,
```
does it cause problems on RHEL7? Can't we just set `$owner = undef`, so that we don't break existing feature?

clehene commented:
@deric I'm not removing them, I'm just setting them globally 
```puppet
  File {
    owner   => $owner,
    group   => $group,
  }
```
(resource default statements - https://docs.puppetlabs.com/puppet/latest/reference/lang_defaults.html) so that we avoid having them set everywhere

deric commented:
Ok, sorry. I've overlooked that.

deric commented on templates/service/systemd/mesos.service.erb:
```diff
@@ -0,0 +1,25 @@
+
+[Service]
+Delegate=yes
+ExecStart=/usr/sbin/mesos-<%= @name %> <% if @name=='slave' -%>-containerizers=docker,mesos<% end -%>
```
If you're using Mesosphere packages it would be easier to use:
```
ExecStart=/usr/bin/mesos-init-wrapper <%= @name %>
```
if you want to execute directly mesos binaries, we have to make sure, that all environment variables are defined.
Either way `-containerizers=docker,mesos` should not be hardcoded.

deric commented:
We should introduce some parameter `configure_via => env_vars` (or `files` for using `/etc/mesos-slave`, `/etc/mesos-master`) which would avoid having same configuration at two places.

deric commented:
This PR is getting too old and some of the issues mentioned are probably already solved, please let me know if any of the issues still persists. I'll try to track the discussion as separate issues:

* [ ] Add support for RHEL7/CentOS7 
* [ ] Minor refactoring to remove (some) duplication in `.pp` files 
* [ ] Added / changed some configs/defaults
* [ ] Forced yum cache expiry to actually get updates
* [ ] Changed configuration templates so that config files could be sourced directly by service units
